4.1 In meinem Modell sind die Skill_moves Daten besonders aussagekräftig weil wenn man die Daten analysiert merkt man das wenn man die Parameter ändert. Wenn man die Grösse und das Gewicht verändert gibt es seltener Veränderungen als wenn man die Skill_moves verändert.

In [2]:
import pandas as pd


Fifa_Datenset = pd.read_csv("Fifa_Datensatz_cleaned.CSV", on_bad_lines= 'skip', sep= ';')

In [3]:
X = Fifa_Datenset[["height_cm", "weight_kgs", "skill_moves(1-5)" ]]
y = Fifa_Datenset[["positions"]]

In [4]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
   X, y, test_size = 0.2, random_state = 42
)

len(X_train), len(X_test)

(14363, 3591)

In [5]:
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import SGDClassifier
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis

algorithms = {
    "Nearest Neighbors" : KNeighborsClassifier(3),
    "Stochastic Gradient Descent" : SGDClassifier(),
    "Linear SVM" : SVC(kernel="linear", C=0.025),
    "Decision Tree" : DecisionTreeClassifier(max_depth=5),
    "Random Forest" : RandomForestClassifier(max_depth=5, n_estimators=10, max_features=1),
    "Neural Net" : MLPClassifier(alpha=1, max_iter=1000),
    "Naive Bayes" : GaussianNB(),
    "LDA" : LinearDiscriminantAnalysis(),
}


for name, algorithm in algorithms.items():

    algorithm.fit(X_train, y_train.values.ravel())

    score = algorithm.score(X_test, y_test)
    print(name, round(score,2))

Nearest Neighbors 0.71
Stochastic Gradient Descent 0.72
Linear SVM 0.75
Decision Tree 0.76
Random Forest 0.76
Neural Net 0.76
Naive Bayes 0.72
LDA 0.75


In [6]:
import joblib
joblib.dump(algorithms["Decision Tree"], 'Fifa_Datenset_decision_tree.joblib')
best_model = joblib.load('Fifa_Datenset_decision_tree.joblib')
best_model

DecisionTreeClassifier(max_depth=5)

Hier unten finden Sie die Aufgaben 4.2/4.3

In [16]:
from sklearn.ensemble import RandomForestRegressor
best_model = RandomForestRegressor(max_depth=45, random_state=0)
best_model.fit(X_train, y_train)
pred = best_model.predict(X_test)

C:\Users\Luis Lüscher\AppData\Local\Temp\ipykernel_24360\3630298034.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  best_model.fit(X_train, y_train)


In [17]:
from sklearn.tree import DecisionTreeClassifier

logreg = DecisionTreeClassifier()

logreg.fit(X_train, y_train)

DecisionTreeClassifier()

In [18]:
y_pred_class = logreg.predict(X_test)

In [19]:
from sklearn import metrics
print(metrics.accuracy_score(y_test, y_pred_class))

0.7529935950988582


In [20]:
print(metrics.confusion_matrix(y_test, y_pred_class))

[[1308  409]
 [ 478 1396]]


In [21]:
confusion = metrics.confusion_matrix(y_test, y_pred_class)
print(confusion)
TP = confusion[1, 1]
TN = confusion[0, 0]
FP = confusion[0, 1]
FN = confusion[1, 0]

[[1308  409]
 [ 478 1396]]


In [22]:
## Das ist meine Messmetrik
from sklearn.metrics import mean_squared_error
mse = mean_squared_error(y_test, pred)
mse

0.1756905328876404

True Positive = 1308, True Negativ =  1396,False Positive = 478, False Negativ = 409


In [23]:
recall = TP/(TP + FN)
precision = TP/(TP + FP)
print(2 * recall * precision / (recall + precision))

0.7589018755096494


In [24]:
##Sensitivität 
sensitivität = TP / float(FN + TP)
(sensitivität)
print(metrics.recall_score(y_test, y_pred_class))

0.7449306296691569


In [25]:
##Spezifität 
spezifität = TN / (TN + FP)
print(spezifität)

0.7617938264414676


4.4 Mein Modell aus Aufgabe 3 funktioniert zu etwa 75 % das heisst das es jedes dritte Mal die Ausgabe richtig ausgibt. Ich denke da Parameter der skill_moves (1-5) am Aussagekräftigsten ist kommt es vor allem auf den Skill_moves Parameter an.  Ich habe nur 75% Prozent die richtige Ausgabe, weil wenn bei Skill_moves 3 mitgegeben wird kommt es nur noch auf das Gewicht und Grösse an und da kommt es teilweise zu Fehlausgaben. Wenn beim Skill_moves Parameter 1,2,4 oder 5 mitgeben wird ist die Vorhersage ziemlich genau.